stdlib/public/runtime/ReflectionMirror.mm

```diff
@@ -0,0 +1,727 @@
+    auto fieldOffset = Struct->getFieldOffsets()[i];
+
+    auto bytes = reinterpret_cast<char*>(value);
+    auto fieldData = reinterpret_cast<OpaqueValue *>(bytes + fieldOffset);
+
+    *outName = getFieldName(Struct->Description->Struct.FieldNames, i);
+    *outFreeFunc = nullptr;
+
+    assert(!fieldType.isIndirect() && "indirect struct fields not implemented");
+    
+    Any result;
+    bool didLoad = loadSpecialReferenceStorage(fieldData, fieldType, &result);
+    if (!didLoad) {
+      result.Type = fieldType.getType();
+      auto *opaqueValueAddr = result.Type->allocateBoxForExistentialIn(&result.Buffer);
+      result.Type->vw_initializeWithCopy(opaqueValueAddr,
+                                         const_cast<OpaqueValue *>(fieldData));
+    }
+    return AnyReturn(result);
+  }
+};
+
+
+// Implementation for enums.
+struct EnumImpl: ReflectionMirrorImpl {
+  bool isReflectable() {
+    const auto Enum = static_cast<const EnumMetadata *>(type);
+    const auto &Description = Enum->Description->Enum;
+
+    // No metadata for C and @objc enums yet
+    if (Description.CaseNames == nullptr)
+      return false;
+
+    return true;
+  }
+  
+  void getInfo(unsigned *tagPtr, const Metadata **payloadTypePtr, bool *indirectPtr) {
+    const auto Enum = static_cast<const EnumMetadata *>(type);
+    const auto &Description = Enum->Description->Enum;
+
+    unsigned payloadCases = Description.getNumPayloadCases();
+
+    // 'tag' is in the range [-ElementsWithPayload..ElementsWithNoPayload-1].
+    int tag = type->vw_getEnumTag(value);
+
+    // Convert resilient tag index to fragile tag index.
+    tag += payloadCases;
+
+    const Metadata *payloadType = nullptr;
+    bool indirect = false;
+
+    if (static_cast<unsigned>(tag) < payloadCases) {
+      auto payload = Description.GetCaseTypes(type)[tag];
+      payloadType = payload.getType();
+      indirect = payload.isIndirect();
+    }
+
+    if (tagPtr)
+      *tagPtr = tag;
+    if (payloadTypePtr)
+      *payloadTypePtr = payloadType;
+    if (indirectPtr)
+      *indirectPtr = indirect;
+  }
+
+  char displayStyle() {
+    return 'e';
+  }
+  
+  intptr_t count() {
+    if (!isReflectable()) {
+      return 0;
+    }
+    
+    const Metadata *payloadType;
+    getInfo(nullptr, &payloadType, nullptr);
+    return (payloadType != nullptr) ? 1 : 0;
+  }
+
+  AnyReturn subscript(intptr_t i, const char **outName,
+                      void (**outFreeFunc)(const char *)) {
+    const auto Enum = static_cast<const EnumMetadata *>(type);
+    const auto &Description = Enum->Description->Enum;
+
+    unsigned tag;
+    const Metadata *payloadType;
+    bool indirect;
+
+    getInfo(&tag, &payloadType, &indirect);
+
+    // Copy the enum payload into a box
+    const Metadata *boxType = (indirect ? &METADATA_SYM(Bo).base : payloadType);
+    BoxPair pair = swift_allocBox(boxType);
+
+    type->vw_destructiveProjectEnumData(const_cast<OpaqueValue *>(value));
```

Comment:
In a +0 context (unless mirrors copy value always), this worries me. I think you may need to perform a full copy of value to be safe.